draft-ietf-gnap-core-protocol.md

```diff
@@ -2428,33 +2428,38 @@ If the AS determines that the request cannot be issued for any
 reason, it responds to the client instance with an error message.
 
 
-error (string)
-: The error code.
+"error" (string)
+:   REQUIRED.  A single ASCII error code from the
+    following, with additional values available in [a registry TBD](#IANA):
 
+    "invalid_request":
+    :     The request is missing a required parameter, includes an
+          invalid parameter value or is otherwise malformed.
 
-~~~
-{
-
-  "error": "user_denied"
-
-}
-~~~
+    "invalid_client":
+    :     The request was made from a client that was not recognized
+          or allowed by the AS, or the client's signature validation failed.
 
+    "user_denied":
+    : The RO denied the request.
 
+    "too_fast":
+    : The client instance did not respect the timeout in the wait response.
 
-The error code is one of the following, with additional values
-available in [a registry TBD](#IANA):
+    "unknown_request":
+    : The request referenced an unknown ongoing access request.
 
+"error_description" (string)
+:   OPTIONAL. A human-readable string description of the error intended for the
+    developer of the client.
 
-user_denied
-: The RO denied the request.
+~~~
+{
 
-too_fast
-: The client instance did not respect the timeout in the
-            wait response.
+  "error": "user_denied"
 
-unknown_request
-: The request referenced an unknown ongoing access request.
+}
+~~~
```

Comment:
I suggest to add:

> Per policy, highly secure AS implementations MAY elect to return `user_denied` upon any error case, in order not to leak any information.

Reply:
I personally don't like this suggestion. `user_denied` should mean that specifically the RO denied the request. If you want to hide the reasons for things failing, there should be a generic `unknown_error` as an option. (Which I'm also not sure about, but like it more than co-opting an otherwise meaningful error code.)

Reply:
@yaronf "access denied" is not the same as "user denied". "User denied" specifically means that the user interactively said "no" to a decision.

Reply:
OK, so maybe we also need a more generic `denied` error code, which means "denied for an unspecified reason".